src/Scout/ScoutEngine.php

```diff
@@ -0,0 +1,535 @@
+     *
+     * @param list<array|object> $results
+     */
+    public function mapIds($results): Collection
+    {
+        assert(is_array($results), new TypeError(sprintf('Argument #1 ($results) must be of type array, %s given', get_debug_type($results))));
+
+        return new Collection(array_column($results, '_id'));
+    }
+
+    /**
+     * Map the given results to instances of the given model.
+     *
+     * @param array $results
+     * @param Model $model
+     */
+    public function map(Builder $builder, $results, $model): Collection
+    {
+        assert(is_array($results), new TypeError(sprintf('Argument #2 ($results) must be of type array, %s given', get_debug_type($results))));
+        assert($model instanceof Model, new TypeError(sprintf('Argument #3 ($model) must be of type %s, %s given', Model::class, get_debug_type($model))));
+
+        if (! $results) {
+            return $model->newCollection();
+        }
+
+        $ids = array_column($results, '_id');
+
+        return $model->getScoutModelsByIds($builder, $ids)
+            ->filter(fn (Model $model) => in_array($model->getScoutKey(), $ids))
+            ->sortBy(fn (Model $model) => array_search($model->getScoutKey(), $ids))
+            ->values();
+    }
+
+    /**
+     * Get the total count from a raw result returned by the engine.
+     * This is an estimate if the count is larger than 1000.
+     *
+     * @param mixed $results
+     */
+    public function getTotalCount($results): int
+    {
+        if (! $results) {
+            return 0;
+        }
+
+        return $results[0]->search_meta['count']['lowerBound'];
+    }
+
+    /**
+     * Flush all records from the engine.
+     *
+     * @param Model $model
+     */
+    public function flush($model): void
+    {
+        assert($model instanceof Model, new TypeError(sprintf('Argument #1 ($model) must be of type %s, %s given', Model::class, get_debug_type($model))));
+
+        $collection = $this->getIndexableCollection($model);
+
+        $collection->deleteMany([]);
+    }
+
+    /**
+     * Map the given results to instances of the given model via a lazy collection.
+     *
+     * @param Builder      $builder
+     * @param array|Cursor $results
+     * @param Model        $model
+     *
+     * @return LazyCollection
+     */
+    public function lazyMap(Builder $builder, $results, $model): LazyCollection
+    {
+        assert($results instanceof Cursor || is_array($results), new TypeError(sprintf('Argument #2 ($results) must be of type %s|array, %s given', Cursor::class, get_debug_type($results))));
+        assert($model instanceof Model, new TypeError(sprintf('Argument #3 ($model) must be of type %s, %s given', Model::class, get_debug_type($model))));
+
+        if (! $results) {
+            return LazyCollection::make($model->newCollection());
+        }
+
+        $objectIds = array_column($results, '_id');
+        $objectIdPositions = array_flip($objectIds);
```

Comment:
I assume this was copied from another engine in the upstream library. Is this the more efficient approach to the `array_search()` algorithm I saw earlier (re: `sortBy()`)?

Reply:
Yes.